File: native-ios/MGLFleetSDK/Sources/MGLFleetSDK/DriverFleetQr.swift
import Foundation

struct FleetpayQrPayloadIOS: Equatable {
    var txnId: String
    var mid: String
    var terminalId: String
    var amountPaise: Int64
    var expiryEpoch: Int64
    var sign: String
    var merchantName: String?
    var currency: String?
}

enum DriverFleetQr {
    static let simulatedUri =
        "fleetpay://pay?txn=SIMTXN01&mid=DEMOMID&tid=DEMOTID&am=67200&exp=9999999999&sign=demosign&mn=Demo%20CNG%20Station"

    static func parseFleetpayPayUri(_ raw: String) -> FleetpayQrPayloadIOS? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard s.lowercased().hasPrefix("fleetpay://") else { return nil }
        guard let u = URLComponents(string: s), let host = u.host?.lowercased() else { return nil }
        let pathNorm = u.path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).lowercased()
        let payPathOk = pathNorm.isEmpty || pathNorm.hasSuffix("pay") || pathNorm.hasSuffix("/pay")
        let payHostOk = host == "pay"
        if !payHostOk && !payPathOk { return nil }

        func qp(_ name: String) -> String? {
            u.queryItems?.first { $0.name == name }?.value?.removingPercentEncoding
                ?? u.queryItems?.first { $0.name == name }?.value
        }

        guard let txnId = qp("txn"), !txnId.isEmpty,
              let mid = qp("mid"), !mid.isEmpty,
              let terminalId = qp("tid"), !terminalId.isEmpty,
              let am = qp("am"), let amountPaise = Double(am).flatMap({ Int64($0) }),
              let exp = qp("exp"), let expiryEpoch = Double(exp).flatMap({ Int64($0) }),
              let sign = qp("sign"), !sign.isEmpty else { return nil }

        return FleetpayQrPayloadIOS(
            txnId: txnId,
            mid: mid,
            terminalId: terminalId,
            amountPaise: amountPaise,
            expiryEpoch: expiryEpoch,
            sign: sign,
            merchantName: qp("mn"),
            currency: qp("cu"),
        )
    }

    static func paiseToInrDisplay(_ amountPaise: Int64) -> String {
        let inr = Double(amountPaise) / 100.0
        let f = NumberFormatter()
        f.locale = Locale(identifier: "en_IN")
        f.minimumFractionDigits = 2
        f.maximumFractionDigits = 2
        return f.string(from: NSNumber(value: inr)) ?? String(format: "%.2f", inr)
    }

    static func validIndianMobile10(_ digitsOnly: String) -> Bool {
        let d = String(digitsOnly.filter(\.isNumber).suffix(10))
        guard d.count == 10, let first = d.first, let fv = Int(String(first)) else { return false }
        return (6 ... 9).contains(fv)
    }

    static func normVrnPublic(_ v: String) -> String {
        v.replacingOccurrences(of: "\\s+", with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespaces)
            .uppercased(with: Locale(identifier: "en_US"))
    }

    static func normVrnForQrPay(_ v: String) -> String {
        normVrnPublic(v).replacingOccurrences(of: " ", with: "")
    }
}

File: native-android/fleet-sdk/src/main/java/com/mgl/fleet/sdk/internal/DriverFleetQr.kt
package com.mgl.fleet.sdk.internal

import java.net.URI
import java.net.URLDecoder

internal fun parseFleetpayPayUri(raw: String): FleetpayQrPayload? {
    val s = raw.trim()
    if (!Regex("^fleetpay://", RegexOption.IGNORE_CASE).containsMatchIn(s)) return null

    val uri =
        try {
            URI.create(s)
        } catch (_: Exception) {
            return null
        }
    val pathNorm = uri.path?.trim('/')?.lowercase()?.trimEnd('/') ?: ""
    val payPathOk = pathNorm.isEmpty() || pathNorm.endsWith("/pay")
    val host = uri.host?.lowercase() ?: ""
    val payHostOk = host == "pay"
    if (!payHostOk && !payPathOk) return null

    fun qp(name: String): String? {
        val q = uri.query ?: return null
        for (part in q.split('&')) {
            val idx = part.indexOf('=')
            val k = if (idx >= 0) part.substring(0, idx) else part
            val v = if (idx >= 0) part.substring(idx + 1) else ""
            if (k == name) {
                return try {
                    URLDecoder.decode(v.replace('+', ' '), Charsets.UTF_8.name())
                } catch (_: Exception) {
                    v
                }
            }
        }
        return null
    }

    val txnId = qp("txn") ?: return null
    val mid = qp("mid") ?: return null
    val terminalId = qp("tid") ?: return null
    val am = qp("am") ?: return null
    val exp = qp("exp") ?: return null
    val sign = qp("sign") ?: return null
    val amountPaise = am.toDoubleOrNull()?.toLong() ?: return null
    val expiryEpoch = exp.toDoubleOrNull()?.toLong() ?: return null
    val mn = qp("mn")

    val merchantName =
        mn?.let { m ->
            try {
                URLDecoder.decode(m.replace('+', ' '), Charsets.UTF_8.name())
            } catch (_: Exception) {
                m
            }
        }

    return FleetpayQrPayload(
        txnId = txnId,
        mid = mid,
        terminalId = terminalId,
        amountPaise = amountPaise,
        expiryEpoch = expiryEpoch,
        sign = sign,
        merchantName = merchantName,
        currency = qp("cu"),
    )
}

internal fun paiseToInrDisplay(amountPaise: Long): String {
    val inr = amountPaise.toDouble() / 100.0
    return String.format(java.util.Locale("en", "IN"), "%.2f", inr)
}

internal fun validIndianMobile10(digitsOnly: String): Boolean {
    val d = digitsOnly.filter(Char::isDigit).takeLast(10)
    if (d.length != 10) return false
    val first = d[0].digitToIntOrNull() ?: return false
    return first in 6..9
}
